src/ripple/app/misc/impl/ValidatorList.cpp

```diff
@@ -0,0 +1,269 @@
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <BeastConfig.h>
+#include <ripple/app/misc/ValidatorList.h>
+#include <ripple/basics/Slice.h>
+#include <ripple/basics/StringUtilities.h>
+#include <ripple/core/ConfigSections.h>
+#include <boost/regex.hpp>
+#include <fstream>
+
+namespace ripple {
+
+static
+PublicKey
+asPublicKey(RippleAddress const& raPublicKey)
+{
+    auto const& blob = raPublicKey.getNodePublic();
+
+    if (blob.empty())
+        LogicError ("Can't convert invalid RippleAddress to PublicKey");
+
+    return PublicKey(Slice(blob.data(), blob.size()));
+}
+
+ValidatorList::ValidatorList (beast::Journal j)
+    : j_ (j)
+{
+}
+
+boost::optional<std::string>
+ValidatorList::member (RippleAddress const& identity) const
+{
+    std::lock_guard <std::mutex> sl (mutex_);
+
+    auto const publicKey = asPublicKey (identity);
+
+    auto ret = ephemeral_.find (publicKey);
+
+    if (ret != ephemeral_.end())
+        return ret->second;
+
+    ret = permanent_.find (publicKey);
+
+    if (ret != permanent_.end())
+        return ret->second;
+
+    return boost::none;
+}
+
+bool
+ValidatorList::trusted (RippleAddress const& identity) const
+{
+    return static_cast<bool> (member(identity));
+}
+
+bool
+ValidatorList::insertEphemeralKey (
+    PublicKey const& identity,
+    std::string comment)
+{
+    std::lock_guard <std::mutex> sl (mutex_);
+
+    if (permanent_.find (identity) != permanent_.end())
+    {
+        JLOG (j_.error) <<
+            toBase58 (TokenType::TOKEN_NODE_PUBLIC, identity) <<
+            ": ephemeral key exists in permanent table!";
+        return false;
+    }
+
+    return ephemeral_.emplace (identity, comment).second;
+}
+
+bool
+ValidatorList::removeEphemeralKey (
+    PublicKey const& identity)
+{
+    std::lock_guard <std::mutex> sl (mutex_);
+    return ephemeral_.erase (identity);
+}
+
+bool
+ValidatorList::insertPermanentKey (
+    RippleAddress const& identity,
+    std::string comment)
+{
+    std::lock_guard <std::mutex> sl (mutex_);
+
+    auto const publicKey = asPublicKey (identity);
+
+    if (ephemeral_.find (publicKey) != ephemeral_.end())
+    {
+        JLOG (j_.error) <<
+            toBase58 (TokenType::TOKEN_NODE_PUBLIC, publicKey) <<
+            ": permanent key exists in ephemeral table!";
+        return false;
+    }
+
+    return permanent_.emplace (publicKey, comment).second;
+}
+
+bool
+ValidatorList::removePermanentKey (
+    RippleAddress const& identity)
+{
+    std::lock_guard <std::mutex> sl (mutex_);
+    return permanent_.erase (asPublicKey (identity));
+}
+
+void
+ValidatorList::for_each (
+    std::function<void(PublicKey const&, std::string const&, bool)> func) const
+{
+    std::lock_guard <std::mutex> sl (mutex_);
+
```

Comment:
These `for_each` functions when the public interface includes functions that could invalidate the collection worry me. It's too easy for a caller to inadvertently invalidate the collection by calling `add` or `remove`.

I don't have a better suggestion, but I wanted to flag this in case other people had good ideas.